Documentation for new QML library (#726)
MicrosoftDocs/quantum-docs-pr: articles/libraries/machine-learning/basic-classification.md

---

> title: Quantum machine learning library

Reviewer: Make the title accurately reflect the intent of the article

description: Learn how to execute a quantum sequential classifier written in Q# using the Quantum Machine Learning library of the Microsoft QDK.
author: geduardo
ms.author: v-edsanc@microsoft.com
ms.date: 02/16/2020
ms.topic: article
uid: microsoft.quantum.libraries.machine-learning.basics
---

# Basic classification: Classify data with the QDK

In this Quickstart, you will learn how to execute a quantum sequential classifier written in Q# using the Quantum Machine Learning library of the QDK. 

In this guide we will use the half-moon dataset, using a classifier structure defined in Q#.

> ## Execute the classifier from the host program

Reviewer: You'll want to include a pre-requisites section here: installation of QDK, set up project etc

> ## Execute the classifier from the host program

Reviewer: ```suggestion
## Host program
```

> First, let's take a look at the host programs to see what the process looks like. Your host program will consist of three parts:

Reviewer: ```suggestion
Your host program consists of three parts:
```

> - Loading of the data from the dataset.

Reviewer: ```suggestion
- Loads the dataset
```

> - Execution of the training to obtain the parameters and the bias that determine the model.

Reviewer: ```suggestion
- Executes training to determine the parameters and bias of the model.
```

> - Validation of the model to test the accuracy.

Reviewer: ```suggestion
- Validates the model to determine its accuracy
```

    ### [Python with Visual Studio Code or the Command Line](#tab/tabid-python)

    To run your the Q# classifier from Python, save the following code as `host.py`. Remember that you also need the Q# file `Training.qs` that is explained later in this tutorial.

    :::code language="python" source="~/quantum/samples/machine-learning/half-moons/host.py" range="3-42":::

    You can then run your Python host program from the command line:

>     ```bash

Reviewer: It wasn't clear to me whether you could run the code at this point or not

Reviewer: If you can't run it, then we should not give instructions to do that

Reviewer: Brief explanation of the starting parameters would be good

Reviewer 2: Yes, true. This was initially written for the Wine sample that didn't have them, but we had to go for Half-moons. Let me add a brief comment.

Reviewer 2: You can run it provided the `Training.qs` file. This file is explained later because it is considerably more complex and daunting than the host programs. I thought it was a good a idea to present the general structure of the process with the host program to have an easier understanding of 'Training.qs' later.

    Preparing Q# environment...
    [...]
    Observed X.XX% misclassifications.
    ```

    ### [C# with Visual Studio Code or the Command Line](#tab/tabid-csharp)

    To run your the Q# classifier from C#, save the following code as `Host.cs`. Remember that you also need the Q# file `Training.qs` that is explained later in this tutorial.

    :::code language="csharp" source="~/quantum/samples/machine-learning/half-moons/Host.cs" range="4-86":::

    You can then run your C# host program from the command line:

    ```bash
    $ dotnet run
    [...]
    Observed X.XX% misclassifications.
    ```

    ### [C# with Visual Studio 2019](#tab/tabid-vs2019)

    To run your new Q# program from C# in Visual Studio, modify `Host.cs` to include the following C# code. Remember that you also need the Q# file `Training.qs` that is explained later in this tutorial.

    :::code language="csharp" source="~/quantum/samples/machine-learning/half-moons/Host.cs" range="4-86":::

    Then press F5, the program will start execution and a new windows will pop-up with the following results: 

    ```bash
    $ dotnet run
    [...]
    Observed X.XX% misclassifications.
    ```
    ***

> ## The Q\# code of the classifier

Reviewer: ```suggestion
## Q\# classifier code
```

> Now let's see how the functions invoked by the host program are defined in Q#. We save the following code in a file named `Training.qs`.

Author: ```suggestion
Now let's see how the operations invoked by the host program are defined in Q#.
We save the following code in a file named `Training.qs`.
```

The most important functions and operations defined in the code above are:

- `ClassifierStructure() : ControlledRotation[]` : in this function we set the structure of our circuit model by adding the layers of the controlled gates we consider. This step is analogous to the declaration of layers of neurons in a sequential deep learning model.
- `TrainHalfMoonModel() : TrainWineModel() : (Double[], Double)` : this operation is the core part of the code and defines the training. Here we load the samples from the dataset included in the library, we set the hyper parameters and the initial parameters for the training and we start the training by calling the operation `TrainSequentialClassifier` included in the library. It outputs the parameters and the bias that determine the classifier.
- `ValidateHalfMoonModel(parameters : Double[], bias : Double) : Int` : this operation defines the validation process to evaluate the model. Here we load the samples for validation, the number of measurements per sample and the tolerance. It outputs the number of misclassifications on the chosen batch of samples for validation.

## Next steps

> First, you can play with the code and try to change some parameters to see how it affects the training. Then, in the next tutorial, [Design your own classifier](xref:microsoft.quantum.libraries.machine-learning.design),  you will learn how to define the structure of the classifier.

Author: Is this a reasonable spot for a next-steps-action div, @natke?
